### DISCUSSION TITLE
<How to reclaimed Docker Image Size>

### DISCUSSION THREAD

**Discussion by suman2007** [community/community#125809] [Programming Help]
### Select Topic Area

Question

### Body

I am creating an Docker image using below docker file.

```
FROM maven:3.8.1-openjdk-11-slim
VOLUME /tmp
ADD . /usr/src/app
WORKDIR /usr/src/app

RUN mkdir -p /DG/activeRelease/IDAP
COPY TestFolder.tar.gz /DGdata/TestFolder.tar.gz

RUN tar -C /DG/activeRelease/IDAP/  -xvf /DGdata/TestFolder.tar.gz &&\
    rm -rf /DGdata/TestFolder.tar.gz

RUN mvn clean package -DskipTests &&\
    tar -C /DG/activeRelease/IDAP/  -xvf /DGdata/TestFolder.tar.gz &&\
    rm -rf /DGdata/TestFolder.tar.gz &&\
    cd /DG/activeRelease/IDAP/TestFolder &&\
     apt-get update &&\
     apt-get -y install iproute2 &&\
     apt-get install -y iputils-ping &&\
     apt-get install -y tcpdump
```
In the above during the execution of "RUN tar -C /DG/activeRelease/IDAP/  -xvf /DGdata/TestFolder.tar.gz &&\
    rm -rf /DGdata/TestFolder.tar.gz" a new space is allocated while unzipping the folder "TestFolder.tar.gz". and now total Image size is 2GB, where for command "RUN tar -C /DG/activeRelease/IDAP/  -xvf /DGdata/TestFolder.tar.gz &&\
    rm -rf /DGdata/TestFolder.tar.gz...." the size is 1GB and for the command "COPY TestFolder.tar.gz /DGdata/TestFolder.tar.gz" the size is 500MB.

To reduce the image size I have changed the Docker file as per below. Now I am directly copying the Folder to avoid unzipping.

```
FROM maven:3.8.1-openjdk-11-slim
VOLUME /tmp
ADD . /usr/src/app
WORKDIR /usr/src/app

RUN mkdir -p /DG/activeRelease/IDAP
COPY TestFolder/  /DGdata/

RUN mvn clean package -DskipTests &&\
    cd /DG/activeRelease/IDAP/TestFolder  &&\
     apt-get update &&\
     apt-get -y install iproute2 &&\
     apt-get install -y iputils-ping &&\
     apt-get install -y tcpdump
```
Although I am getting the same Image size i.e. 2GB and in Layer
RUN mvn clean package -DskipTests &&\
    cd /DG/activeRelease/IDAP/TestFolder  &&\
     ........
which is 1GB.

My question is why does space is not reduced by the size of (TestFolder.tar.gz) which is 500MB.
Where the unzip size of the folder is 750MB.
So after that the total size must be (2GB - 500MB) = 1.5GB

**Comment by judahpaul16** [comment]
Each command in the Dockerfile creates a new layer. When you copy the `TestFolder.tar.gz` and then expand it, these actions take place in separate layers:

1. The `COPY TestFolder.tar.gz /DGdata/TestFolder.tar.gz` command creates a layer with the `.tar.gz` file.
2. The `RUN tar -C /DG/activeRelease/IDAP/ -xvf /DGdata/TestFolder.tar.gz && rm -rf /DGdata/TestFolder.tar.gz` command creates another layer where the tar file is expanded and then the original tar.gz file is deleted. Even though the tar.gz file is deleted in this layer, the space it occupied in the previous layer still exists.

In your attempt to reduce the size by directly copying the folder instead of the `tar.gz` file, you're essentially replacing the `tar.gz` layer with a similar size (or possibly larger, depending on the compression ratio) folder layer. The size of the `TestFolder/` might be around or larger than the compressed `TestFolder.tar.gz` because compression can significantly reduce file size. Each layer is immutable. Even if files are deleted in subsequent layers, the space they occupied in earlier layers remains consumed because those earlier layers are unchanged.

My advice:

- Combine commands into a single RUN instruction where possible. This reduces the number of layers and can help in reducing the overall size since all operations are done in a single filesystem layer.
- Whenever installing packages or doing temporary operations, make sure to clean up within the same RUN instruction to avoid bloating the image with unnecessary data.
- Build your application in a "builder" stage with all the necessary tools and dependencies, and then copy only the necessary artifacts (like your compiled application) to a final image. This can significantly reduce the image size as only the necessary files are included in the final image.

Maybe something like this:
```dockerfile
# Builder stage
FROM maven:3.8.1-openjdk-11-slim as builder
WORKDIR /app
ADD . /app
RUN mvn clean package -DskipTests

# App stage
FROM openjdk:11-slim
COPY --from=builder /app/target/my-app.jar /app/
WORKDIR /app
RUN apt-get update &&\
    apt-get install -y iproute2 iputils-ping tcpdump &&\
    apt-get clean &&\
    rm -rf /var/lib/apt/lists/*
CMD ["java", "-jar", "my-app.jar"]
```

**Reply by suman2007** [reply]
Thanks @judahpaul16 
It's true and I agree with your point.
Please correct me if I am wrong.
Layer-1 
`COPY TestFolder.tar.gz /DGdata/TestFolder.tar.gz`   --> New space allocated as per the .tar e.g. 500MB

Layer-2 
`tar -C /DG/activeRelease/IDAP/  -xvf /DGdata/TestFolder.tar.gz` --> Another space allocated during untar and to hold the untar folder e.g. 750MB

**Reply by judahpaul16** [reply]
Well technically the first layer is FROM and also ADD creates a layer with all the files in your build context (directory with the dockerfile) so COPY may be redundantly adding size to the image.

**Reply by suman2007** [reply]
Thanks,
Layer-1
COPY TestFolder/ /DGdata/ --> New space allocated as per the .tar e.g. 750MB

If I copy the whole folder it will create a chunk of 750MB due to the size of "TestFolder".
Then I don't need to untar in Layer-2.
So, I guess the overall size will be decreased by 500 MB, but actually, it's not happening. I am curious to know why it's not happening. The overall size is same

**Reply by judahpaul16** [reply]
Keep in mind `ADD . /usr/src/app` is basically a fancy `COPY` so if the extracted `TestFolder/` or `tar.gz` files are in your build context then the `ADD` layer will include them in the final image. You could check this by seeing if `TestFolder/` or the `tar.gz` exists in `/usr/src/app` after the the container is built.

You could also inspect the size of each layer with:
```bash
docker history --no-trunc --format "{{.ID}}: {{.CreatedBy}} -> {{.Size}}" [IMAGE]
```

**Reply by suman2007** [reply]
You are absolutely right. The tar.gz is present in /usr/src/app
ADD . /usr/src/app is constant for each Dockerfile.
Total image size includes the size of /usr/src/app.

The point is for 
1)----->
```
RUN mkdir -p /DG/activeRelease/IDAP
COPY TestFolder.tar.gz /DGdata/TestFolder.tar.gz
```
2)---->
```
RUN mkdir -p /DG/activeRelease/IDAP
COPY TestFolder/  /DGdata/

RUN mvn clean package -DskipTests &&\
    cd /DG/activeRelease/IDAP/TestFolder
```

How is in the above cases the total Image size is the same?

**Reply by judahpaul16** [reply]
Was your build context exactly the same in both cases? In other words, did both scenarios include the extracted folder and the `tar.gz` in the dockerfile directory. If in the first scenario there is a `tar.gz` (500mb) and no `TestFolder/` (750mb) then the final image will be +1250mb because the tar is being added and extracted. If in the second scenario there is only the folder and no `tar.gz` in the build context then the final image will be +1500mb because the folder is being added and copied.

**Comment by suman2007** [comment]
After a long break....
I have observation...
I have cloud where I placed the image, but when pulled in local the size increases almost 100MB.
Will docker image size increases when is it being pulled?
